File: executors/claude-agent/src/observability.ts
// Copyright © 2026 Fall Guy Consulting.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.apache at the repo root.

// observability.ts — claude-agent's ExecutorObservability surface.
//
// Records per-dispatch trace events in a bounded in-memory ledger.
// Exposes:
//   - GET /observability/v1/capabilities
//   - GET /observability/v1/trace/{dispatch_id}
//   - GET /observability/v1/trace/{dispatch_id}/stream  (SSE)
//
// Per spec §2: standard vocabulary + free-form fallback. The agent
// emits step_started/step_completed/step_failed/tool_call/error/log
// via `recordEvent`; consumers see them via getTrace / streamTrace.
//
// gRPC service registration is deferred to v2 — the dashboard's
// Hono proxy talks HTTP+JSON regardless, and the conformance probe
// in `core/cmd/rimsky-conformance/` exercises the HTTP+JSON surface.
//
// Bounded retention: per dispatch, all events are retained until
// retention_after_terminal_seconds elapses past terminal; then evicted
// (the trace returns evicted: true, complete: true, events: []).

import { randomUUID } from "node:crypto";
import type { FastifyInstance } from "fastify";

export type Severity = "DEBUG" | "INFO" | "WARN" | "ERROR";

export interface TraceEvent {
  event_id: string;
  parent_event_id?: string;
  timestamp: string;
  severity: Severity;
  category: string;
  message?: string;
  attributes?: Record<string, unknown>;
}

export interface Trace {
  dispatch_id: string;
  evicted: boolean;
  complete: boolean;
  events: TraceEvent[];
}

interface TraceRecord {
  dispatch_id: string;
  events: TraceEvent[];
  complete: boolean;
  terminalAt?: number;
  // listeners for live SSE streams
  listeners: Set<(ev: TraceEvent) => void>;
}

const RETENTION_AFTER_TERMINAL_SECONDS = 3600;
const MAX_TRACES = 1024;

/**
 * @agent-contract
 * what: Per-dispatch in-memory trace ledger backing the executor
 *   observability protocol (HTTP+JSON bridge mounted by `mountObservability`).
 * how: `recordEvent(dispatchId, ev)` appends one event; `markComplete(dispatchId)`
 *   stamps the terminal time; `getTrace(dispatchId)` returns a snapshot.
 *   Tests construct a fresh `Observability` per case; the production singleton
 *   is created by `main.ts` and shared with `agent-run.ts` via a callback.
 * handles: bounded retention (eviction at retention + idle check), free-form
 *   categories, SSE listener fan-out.
 * does-not-handle: persistence (process-local only); cross-process replication;
 *   gRPC service hosting (HTTP+JSON only in v1).
 */
export class Observability {
  private records = new Map<string, TraceRecord>();
  private order: string[] = [];

  recordEvent(dispatchId: string, ev: Partial<TraceEvent> & Pick<TraceEvent, "category">) {
    const rec = this.getOrCreate(dispatchId);
    const event: TraceEvent = {
      event_id: ev.event_id ?? randomUUID(),
      parent_event_id: ev.parent_event_id,
      timestamp: ev.timestamp ?? new Date().toISOString(),
      severity: ev.severity ?? "INFO",
      category: ev.category,
      message: ev.message,
      attributes: ev.attributes,
    };
    rec.events.push(event);
    for (const cb of rec.listeners) {
      try {
        cb(event);
      } catch {
        // listener failures are not the producer's problem
      }
    }
    this.evictIfNeeded();
  }

  markComplete(dispatchId: string) {
    const rec = this.getOrCreate(dispatchId);
    if (rec.complete) return;
    rec.complete = true;
    rec.terminalAt = Date.now();
    const tail: TraceEvent = {
      event_id: randomUUID(),
      timestamp: new Date().toISOString(),
      severity: "INFO",
      category: "trace_complete",
    };
    rec.events.push(tail);
    for (const cb of rec.listeners) {
      try {
        cb(tail);
      } catch {
        // ignore
      }
    }
  }

  getTrace(dispatchId: string): Trace {
    const rec = this.records.get(dispatchId);
    if (!rec || this.isEvicted(rec)) {
      // Per spec §2.6: missing dispatches must surface as the same
      // evicted-shape envelope. evicted:true makes "we don't have
      // it" a single observable signal, regardless of whether the
      // dispatch never existed or has been evicted by retention.
      return { dispatch_id: dispatchId, evicted: true, complete: true, events: [] };
    }
    return {
      dispatch_id: dispatchId,
      evicted: false,
      complete: rec.complete,
      events: [...rec.events],
    };
  }

  /**
   * Subscribe to live events for a dispatch. Returns an unsubscribe
   * function. Replay is the caller's concern (the HTTP handler emits
   * the snapshot first, then attaches the listener).
   *
   * Note: prefer `subscribeWithSnapshot` for SSE handlers — it
   * atomically returns the current snapshot + attaches the listener
   * under one async tick, eliminating the gap window where new events
   * could land between snapshot capture and listener registration
   * (issue 11). Plain `subscribe` is kept for tests and for callers
   * that don't need the snapshot.
   */
  subscribe(dispatchId: string, cb: (ev: TraceEvent) => void): () => void {
    const rec = this.getOrCreate(dispatchId);
    rec.listeners.add(cb);
    return () => {
      rec.listeners.delete(cb);
    };
  }

  /**
   * Atomic snapshot+subscribe. Returns the current event slice plus
   * a teardown handle for the live listener. Because both reads/writes
   * happen synchronously in JS's single-threaded event loop, no event
   * appended elsewhere can land between the snapshot and the listener
   * attach — this is the analog of the Go-side fix in issue 11.
   */
  subscribeWithSnapshot(
    dispatchId: string,
    cb: (ev: TraceEvent) => void,
  ): { snapshot: Trace; unsubscribe: () => void } {
    const rec = this.getOrCreate(dispatchId);
    if (this.isEvicted(rec)) {
      return {
        snapshot: { dispatch_id: dispatchId, evicted: true, complete: true, events: [] },
        unsubscribe: () => {},
      };
    }
    const snapshot: Trace = {
      dispatch_id: dispatchId,
      evicted: false,
      complete: rec.complete,
      events: [...rec.events],
    };
    rec.listeners.add(cb);
    return {
      snapshot,
      unsubscribe: () => {
        rec.listeners.delete(cb);
      },
    };
  }

  private getOrCreate(dispatchId: string): TraceRecord {
    let rec = this.records.get(dispatchId);
    if (rec) return rec;
    rec = { dispatch_id: dispatchId, events: [], complete: false, listeners: new Set() };
    this.records.set(dispatchId, rec);
    this.order.push(dispatchId);
    this.evictIfNeeded();
    return rec;
  }

  private isEvicted(rec: TraceRecord): boolean {
    if (!rec.complete || !rec.terminalAt) return false;
    return Date.now() - rec.terminalAt > RETENTION_AFTER_TERMINAL_SECONDS * 1000;
  }

  private evictIfNeeded() {
    // Hard cap on map size: when the bound is exceeded, drop the
    // oldest record regardless of state. Without this, a long run of
    // never-terminal dispatches would grow the ledger unbounded
    // (parallels the same fix on the Go ledgers).
    while (this.records.size > MAX_TRACES) {
      const oldestId = this.order.shift();
      if (!oldestId) break;
      this.records.delete(oldestId);
    }
  }
}

export function capabilitiesPayload(httpBridgeUrl = "") {
  return {
    supports_trace_get: true,
    supports_trace_stream: true,
    retention_after_terminal_seconds: RETENTION_AFTER_TERMINAL_SECONDS,
    custom_ui: null,
    http_bridge_url: httpBridgeUrl,
  };
}

/**
 * Mounts the observability HTTP routes onto an existing Fastify
 * instance. The same routes are added by `startObservabilityServer`
 * when the executor boots its observability surface as a standalone
 * server; alternatively a parent listener (e.g. the http-bridge) can
 * mount them under itself.
 *
 * httpBridgeUrl, when non-empty, is included in the capabilities
 * response so dashboards can discover the externally-reachable URL
 * for browser-friendly fetch/SSE.
 */
export function mountObservability(
  app: FastifyInstance,
  obs: Observability,
  httpBridgeUrl = "",
): void {
  app.get("/observability/v1/capabilities", async () =>
    capabilitiesPayload(httpBridgeUrl),
  );

  app.get<{ Params: { dispatch_id: string } }>(
    "/observability/v1/trace/:dispatch_id",
    async (req) => obs.getTrace(req.params.dispatch_id),
  );

  app.get<{ Params: { dispatch_id: string } }>(
    "/observability/v1/trace/:dispatch_id/stream",
    async (req, reply) => {
      const dispatchId = req.params.dispatch_id;
      reply
        .header("Content-Type", "text/event-stream")
        .header("Cache-Control", "no-cache")
        .header("Connection", "keep-alive");
      const send = (ev: TraceEvent) => {
        reply.raw.write(`data: ${JSON.stringify(ev)}\n\n`);
      };
      // Atomic snapshot+subscribe so events appended between the two
      // can't escape the stream (issue 11).
      const result = obs.subscribeWithSnapshot(dispatchId, (ev) => {
        send(ev);
        if (ev.category === "trace_complete") {
          result.unsubscribe();
          reply.raw.end();
        }
      });
      for (const ev of result.snapshot.events) send(ev);
      if (result.snapshot.evicted || result.snapshot.complete) {
        result.unsubscribe();
        reply.raw.end();
        return;
      }
      // Spec §2.5: idle-close after RIMSKY_OBS_IDLE_TIMEOUT_MS (default
      // 5 minutes). The listener resets the timer on every event;
      // disconnect cancels both timer and subscription.
      const idleMs = Number(process.env.RIMSKY_OBS_IDLE_TIMEOUT_MS ?? 5 * 60 * 1000);
      let idleTimer: ReturnType<typeof setTimeout> | null = null;
      const armIdle = () => {
        if (idleMs <= 0) return;
        if (idleTimer !== null) clearTimeout(idleTimer);
        idleTimer = setTimeout(() => {
          // Final keepalive comment + close (not an error).
          reply.raw.write(`: idle_timeout\n\n`);
          result.unsubscribe();
          reply.raw.end();
        }, idleMs);
      };
      armIdle();
      req.raw.on("close", () => {
        if (idleTimer !== null) clearTimeout(idleTimer);
        result.unsubscribe();
      });
    },
  );
}

File: executors/http-node/observability.go
// Copyright © 2026 Fall Guy Consulting.
// Licensed under the Apache License, Version 2.0. See LICENSE.apache at the
// repo root, or http://www.apache.org/licenses/LICENSE-2.0.

package main

import (
	"context"
	"sync"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/structpb"
	"google.golang.org/protobuf/types/known/timestamppb"

	genv1 "github.com/fallguy/rimsky/protocols/proto/v1/gen"
)

// retentionSeconds is the default trace retention window after a
// dispatch terminals. Exposed in capabilities; the eviction sweep
// deletes traces older than this.
const retentionSeconds uint64 = 3600

// defaultStreamIdleTimeout is the spec §2.5 close-idle default for
// StreamTrace listeners with no recent event traffic.
const defaultStreamIdleTimeout = 5 * time.Minute

// traceRecord holds the events for one dispatch plus terminal state.
//
// Note on @blessed-invariant 11: that invariant scopes Rimsky core's
// behavior toward the userdata field on the wire. Executor-supplied
// trace `attributes` are produced by the executor itself and are
// fully introspectable by the executor — invariant 11 does NOT
// restrict an executor's freedom to inspect, structure, or expose
// its own trace data.
type traceRecord struct {
	events     []*genv1.TraceEvent
	terminal   bool
	terminalAt time.Time
	registered bool // true when a dispatch hook has formally claimed this id
}

// subscriber represents one live StreamTrace listener for a dispatch.
// Per spec §2.6 events are never dropped: subscribers read events out
// of the per-dispatch slice (rec.events) at their own cursor (cursor),
// woken up by a buffered wakeup channel each time AppendEvent appends
// or MarkTerminal closes the subscription. AppendEvent never blocks
// on a slow consumer because the wakeup channel is non-blocking
// (capacity-1, coalescing pending wakeups), and there's no per-event
// channel send.
type subscriber struct {
	wake chan struct{} // capacity 1; coalesces multiple appends
	done chan struct{} // closed by MarkTerminal to signal terminal
}

// ObservabilityServer is the http-node observability impl. Traces are
// kept in a per-dispatch in-memory ring with a simple TTL sweep. Live
// streaming is implemented via per-dispatch wakeup-pumped subscribers
// reading directly out of the trace slice — no per-event channel send,
// so AppendEvent never drops events under contention.
type ObservabilityServer struct {
	genv1.UnimplementedExecutorObservabilityServer

	mu          sync.RWMutex
	traces      map[string]*traceRecord
	subs        map[string]map[*subscriber]struct{}
	idleTimeout time.Duration
	// httpBridgeURL is set once at startup before the gRPC server
	// accepts traffic; using sync.Once makes that contract loud.
	httpBridgeURLOnce sync.Once
	httpBridgeURL     string
}

// NewObservabilityServer constructs an empty in-memory trace store
// guarded by sync.RWMutex.
func NewObservabilityServer() *ObservabilityServer {
	return &ObservabilityServer{
		traces:      map[string]*traceRecord{},
		subs:        map[string]map[*subscriber]struct{}{},
		idleTimeout: defaultStreamIdleTimeout,
	}
}

// SetHTTPBridgeURL records the URL the executor advertises in
// ObservabilityCapabilities.http_bridge_url. Set-once at startup;
// subsequent calls are ignored. Empty value disables the hint.
func (s *ObservabilityServer) SetHTTPBridgeURL(u string) {
	s.httpBridgeURLOnce.Do(func() {
		s.mu.Lock()
		defer s.mu.Unlock()
		s.httpBridgeURL = u
	})
}

// SetIdleTimeout overrides the default StreamTrace idle close timeout.
// Pass zero to disable the timeout. Must be set before any stream
// subscribes (set-once-at-startup).
func (s *ObservabilityServer) SetIdleTimeout(d time.Duration) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.idleTimeout = d
}

// RegisterDispatch is the explicit dispatch-hook entry point. Creates
// the per-dispatch trace record so subsequent AppendEvent / MarkTerminal
// calls succeed without auto-creating records for forged ids.
//
// Per issue 13: AppendEvent and MarkTerminal now require the dispatch
// to be registered, so a fabricated dispatch_id can no longer fill the
// in-memory ledger.
func (s *ObservabilityServer) RegisterDispatch(dispatchID string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	rec, ok := s.traces[dispatchID]
	if !ok {
		rec = &traceRecord{}
		s.traces[dispatchID] = rec
	}
	rec.registered = true
}

// Capabilities reports the http-node observability surface:
// supports both GetTrace and StreamTrace, retention 1 hour, no custom
// UI.
func (s *ObservabilityServer) Capabilities(_ context.Context, _ *genv1.ExecutorCapabilitiesRequest) (*genv1.ObservabilityCapabilities, error) {
	s.mu.RLock()
	url := s.httpBridgeURL
	s.mu.RUnlock()
	return &genv1.ObservabilityCapabilities{
		SupportsTraceGet:              true,
		SupportsTraceStream:           true,
		RetentionAfterTerminalSeconds: retentionSeconds,
		HttpBridgeUrl:                 url,
	}, nil
}

// GetTrace returns the snapshot for dispatch_id. When the trace is
// past retention or evicted, returns an empty Trace with evicted=true.
func (s *ObservabilityServer) GetTrace(_ context.Context, req *genv1.GetTraceRequest) (*genv1.Trace, error) {
	if req.GetDispatchId() == "" {
		return nil, status.Error(codes.InvalidArgument, "dispatch_id required")
	}
	s.mu.RLock()
	defer s.mu.RUnlock()
	rec, ok := s.traces[req.GetDispatchId()]
	if !ok {
		return &genv1.Trace{DispatchId: req.GetDispatchId(), Evicted: true, Complete: true}, nil
	}
	out := make([]*genv1.TraceEvent, len(rec.events))
	copy(out, rec.events)
	return &genv1.Trace{
		DispatchId: req.GetDispatchId(),
		Complete:   rec.terminal,
		Events:     out,
	}, nil
}

// StreamTrace replays the snapshot then streams new events; closes
// with a synthetic trace_complete event when the dispatch terminals.
// Unknown / evicted dispatches are reported via the same evicted-shape
// trace_complete marker GetTrace returns (spec §2.6).
//
// Implementation: the subscriber is registered atomically with the
// initial cursor (0), so AppendEvent's wakeup signal always arrives
// after the cursor has been recorded. The pump loop reads events out
// of rec.events at its cursor under a short read-lock, so no events
// can fall in a gap window between snapshot and live attach.
func (s *ObservabilityServer) StreamTrace(req *genv1.StreamTraceRequest, stream genv1.ExecutorObservability_StreamTraceServer) error {
	if req.GetDispatchId() == "" {
		return status.Error(codes.InvalidArgument, "dispatch_id required")
	}
	dispatchID := req.GetDispatchId()
	sub, exists := s.subscribe(dispatchID)
	if !exists {
		// Mirror GetTrace's evicted shape: emit one trace_complete
		// marker and close cleanly. Per spec §2.6 the snapshot and
		// stream surfaces must agree on missing-dispatch behavior.
		return stream.Send(traceCompleteEvent())
	}
	defer s.unsubscribe(dispatchID, sub)
	cursor := 0
	s.mu.RLock()
	idle := s.idleTimeout
	s.mu.RUnlock()
	for {
		events, terminal := s.drainFrom(dispatchID, cursor)
		cursor += len(events)
		for _, ev := range events {
			if err := stream.Send(ev); err != nil {
				return err
			}
		}
		if terminal {
			return stream.Send(traceCompleteEvent())
		}
		var idleC <-chan time.Time
		if idle > 0 {
			t := time.NewTimer(idle)
			idleC = t.C
			defer t.Stop()
		}
		select {
		case <-stream.Context().Done():
			return nil
		case <-sub.done:
			// Drain any tail events MarkTerminal made visible before
			// closing the subscription, then emit trace_complete.
			tail, _ := s.drainFrom(dispatchID, cursor)
			for _, ev := range tail {
				if err := stream.Send(ev); err != nil {
					return err
				}
			}
			return stream.Send(traceCompleteEvent())
		case <-idleC:
			// Spec §2.5: close idle streams cleanly with a final
			// keepalive marker, not an error.
			return stream.Send(traceCompleteEvent())
		case <-sub.wake:
			// Loop back and drain.
		}
	}
}

// subscribe registers a new subscriber for the dispatch's live event
// stream. Returns (sub, false) when the dispatch is unknown to the
// ledger — callers handle that case by emitting an evicted-shape
// terminal marker. The cursor starts at 0; the StreamTrace loop reads
// events out of the trace's own slice, so atomicity comes from the
// fact that AppendEvent's wakeup signal is sent under the same lock
// that appended the event — no gap window.
func (s *ObservabilityServer) subscribe(dispatchID string) (*subscriber, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, ok := s.traces[dispatchID]; !ok {
		return nil, false
	}
	sub := &subscriber{
		wake: make(chan struct{}, 1),
		done: make(chan struct{}),
	}
	if s.subs[dispatchID] == nil {
		s.subs[dispatchID] = map[*subscriber]struct{}{}
	}
	s.subs[dispatchID][sub] = struct{}{}
	return sub, true
}

// unsubscribe drops sub from the dispatch's subscriber set. Idempotent.
func (s *ObservabilityServer) unsubscribe(dispatchID string, sub *subscriber) {
	if sub == nil {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if subs, ok := s.subs[dispatchID]; ok {
		delete(subs, sub)
	}
}

// drainFrom returns a copy of the dispatch's events from cursor onward
// plus the terminal flag. Caller iterates the snapshot lock-free.
func (s *ObservabilityServer) drainFrom(dispatchID string, cursor int) ([]*genv1.TraceEvent, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	rec, ok := s.traces[dispatchID]
	if !ok {
		return nil, true
	}
	if cursor >= len(rec.events) {
		return nil, rec.terminal
	}
	out := make([]*genv1.TraceEvent, len(rec.events)-cursor)
	copy(out, rec.events[cursor:])
	return out, rec.terminal
}

// AppendEvent adds an event to dispatchID's trace and wakes any live
// StreamTrace subscribers. Called by executor dispatch-flow hooks.
//
// Per spec §2.6 events are never dropped: the subscriber pump reads
// directly out of the per-dispatch event slice at its own cursor, so
// the only thing the broadcast needs to do is signal that new events
// are available. The wake channel is capacity-1, so a non-blocking
// send is correct — multiple pending wakes coalesce into one drain
// pass on the receiver side.
//
// Returns silently when the dispatch has not been formally registered
// via RegisterDispatch: forged or unknown dispatch IDs cannot fill the
// in-memory ledger via the executor's bridge surface.
func (s *ObservabilityServer) AppendEvent(dispatchID string, ev *genv1.TraceEvent) {
	s.mu.Lock()
	defer s.mu.Unlock()
	rec, ok := s.traces[dispatchID]
	if !ok || !rec.registered {
		return
	}
	rec.events = append(rec.events, ev)
	for sub := range s.subs[dispatchID] {
		select {
		case sub.wake <- struct{}{}:
		default:
		}
	}
}

// MarkTerminal stamps a dispatch as terminal and signals every live
// subscriber. Subscribers will drain the final tail of events out of
// rec.events and then emit trace_complete. Returns silently for
// unregistered dispatch IDs (no auto-creation; see issue 13).
//
// terminalAt is captured exactly once on the first MarkTerminal call;
// subsequent calls are no-ops, so trace_complete timestamps stay
// stable across follow-on GetTrace requests.
func (s *ObservabilityServer) MarkTerminal(dispatchID string) {
	s.mu.Lock()
	rec, ok := s.traces[dispatchID]
	if !ok || !rec.registered {
		s.mu.Unlock()
		return
	}
	if rec.terminal {
		// Already terminal — keep the original terminalAt; no-op.
		s.mu.Unlock()
		return
	}
	rec.terminal = true
	rec.terminalAt = time.Now()
	subs := s.subs[dispatchID]
	delete(s.subs, dispatchID)
	s.mu.Unlock()
	for sub := range subs {
		close(sub.done)
	}
}

// SweepEvicted removes traces whose terminal timestamp exceeds the
// retention window. Called periodically by the dispatch loop or a
// sweeper goroutine.
func (s *ObservabilityServer) SweepEvicted(now time.Time) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for id, rec := range s.traces {
		if rec.terminal && now.Sub(rec.terminalAt) > time.Duration(retentionSeconds)*time.Second {
			delete(s.traces, id)
		}
	}
}

func traceCompleteEvent() *genv1.TraceEvent {
	return &genv1.TraceEvent{
		EventId:   "trace_complete",
		Timestamp: timestamppb.Now(),
		Severity:  genv1.Severity_INFO,
		Category:  "trace_complete",
	}
}

// MakeEvent is a small constructor used by the dispatch hooks to emit
// a trace event. attrs may be nil.
func MakeEvent(eventID, parentID, category, message string, sev genv1.Severity, attrs map[string]any) *genv1.TraceEvent {
	var pb *structpb.Struct
	if attrs != nil {
		pb, _ = structpb.NewStruct(attrs)
	}
	return &genv1.TraceEvent{
		EventId:       eventID,
		ParentEventId: parentID,
		Timestamp:     timestamppb.Now(),
		Severity:      sev,
		Category:      category,
		Message:       message,
		Attributes:    pb,
	}
}

// RegisterObservability registers the http-node observability server on
// srv. Returns the server so callers can wire it into dispatch hooks.
func RegisterObservability(srv *grpc.Server) *ObservabilityServer {
	o := NewObservabilityServer()
	genv1.RegisterExecutorObservabilityServer(srv, o)
	return o
}
